### DISCUSSION TITLE
POLL - Which Java version are you using?

### DISCUSSION THREAD

**Discussion by luigidellaquila** [orientechnologies/orientdb#9426] [General]
Hi all,

You probably know that in Java 15 a few scripting APIs were removed (see https://openjdk.java.net/jeps/372).
This is a big change by itself, but it also changes our perspective for the future in terms of backward compatibility.

You know OrientDB can also run embedded (in-process), so every time we make a change in terms of JVM requirements, we risk to make it impossible to run it inside a YOUR application (because of ITS requirements!). This is why we always tried to keep it compatible with old JVM versions (ie. Java 8), relying on the fact that recent JVMs guaranteed to be able to run old bytecode.

The above does not apply anymore!

Now we have to make some choices:
- we can stick to Java 8, but you won't be able to embed OrientDB in a Java 15 JVM
- we can upgrade to Java 15, it will give us a chance to improve OrientDB in many aspects, but you won't be able to run it in previous versions (eg. Java 8) anymore
- we could also decide to maintain TWO different branches, one for old JVMs, one for the new ones, but iti will increase the effort a lot

We are already analysing the impact of replacing Nashorn scripting engine with GraalVM, but it is not enough to solve all the problems; we also have to **make a choice about the JVM compatibility**.

**To do this, we need your help**

We like to take informed decisions, these decisions in particular will have a big impact on YOUR operations, so we need to know what **you really need**.

### If you have one minute, please answer the following few questions:

- Do you run OrientDB embedded or stand-alone?
- Which JVM version are you running?
- Are you planning to upgrade JVM in the near future? To which version?
- Do you have constraints about this upgrade? 

If you have any further observations or questions, please do not hesitate.

Thank you!

Luigi

**Comment by andrii0lomakin** [comment]
HI @luigidellaquila ,
I suppose the list of choices which you listed is a bit limited.

For example, it will be still possible to use ODB with Java 15 if you do not use JavaScript-based functions. Which unlikely problem for the embedded distributions.

But even taking embeddability question aside. There are other options. I see the usage of GraalVM as vendor lock, unfortunately. Let's remember the history of commercialization of Oracle JVMs since JDK 8+ after which many projects switched to OpenJDK distributions. What prevents Oracle to change GraalVM license and ask money for it. We already have seen such cases in the past. 

As for me feasible to consider other variants too. For example deprecation of JavaScript functions and migration to Python-based functions.  So support of JavaScript can be dropped for the users which start to use JVM 15 while other users will still enjoy if they wish, JavaScript-based functions. Users even can do such deprecation themselves if JavaScript/Pyhton/etc. support will be implemented as modules.

**Reply by luigidellaquila** [reply]
Hi @laa 

Sure, there are other alternatives. 
Still, we need to understand the community needs to take informed decisions. If 100% of the user base is using Java 15 (it's not the case of course, just an extreme example), maintaining Nashorn-based implementation is just pointless...

Thanks

Luigi

**Reply by andrii0lomakin** [reply]
Yes, I am very curious too.

**Comment by mcgarnagle** [comment]
**Do you run OrientDB embedded or stand-alone?**
We run stand-alone.

**Which JVM version are you running?**
JVM 8

**Are you planning to upgrade JVM in the near future? To which version?**
Yes planning to upgrade in the near future to JVM 15

**Do you have constraints about this upgrade?**
None.

From a selfish point of view, I'd like to see you upgrade to JVM 15 so you can make improvements to OrientDB.

**Comment by zerovian** [comment]
we use java 11. we use it embedded in a web application. no plans to go to java 15 because it is not an LTS version.  We will likely adopt the next java LTS version sometime after it comes out.

java 8 is not a concern at all as we stopped supporting it 2 years ago.

We don't use the javascript engine (in fact we had to turn it off because AIX did not ship with one).

**Comment by scrabb** [comment]
**Do you run OrientDB embedded or stand-alone?**
stand-alone.

**Which JVM version are you running?**
JVM 13

**Are you planning to upgrade JVM in the near future? To which version?**
Yes JVM 15

**Do you have constraints about this upgrade?**
None.

**Comment by mmariuzzo** [comment]
**Do you run OrientDB embedded or stand-alone?**
stand-alone.

**Which JVM version are you running?**
JVM 8 for OrientDB and for all the java based products we use in our solutions

**Are you planning to upgrade JVM in the near future? To which version?**
We'll stay on jvm 8 in the near future. If we have to move, we'll move on a recent LTS version, but depends on what's happen in the products we use.

Do you have constraints about this upgrade?
None. 

We don't use javascript functions, but have it (or python) as an optional plugin could be a good solution

**Comment by johny65** [comment]
**Do you run OrientDB embedded or stand-alone?**
Stand-alone.

**Which JVM version are you running?**
Java 11.

**Are you planning to upgrade JVM in the near future? To which version?**
Yes, next LTS.

**Do you have constraints about this upgrade?**
We have a Java agent that we must adapt to run in the new version.

**Comment by odysseaspenta** [comment]
**Do you run OrientDB embedded or stand-alone?**
We use OrientDB in embedded mode primarily and considering remote mode for future deployments.

**Which JVM version are you running?**
We are using JVM 8 with OrientDB 2.2.x and migrating to JVM 11 with 3.0.x

**Are you planning to upgrade JVM in the near future? To which version?**
We plan to stay with JVM 11 for now and eventually migrate to the next LTS

**Do you have constraints about this upgrade?**
We do use Javascript functions but if you provide a Python alternative that is documented, we can switch to it

**Comment by mdre** [comment]
Hi Luigi!

Here is our stats:

Do you run OrientDB embedded or stand-alone?
We run it in stand-alone mode.

Which JVM version are you running?
All of our projects has been moved to OpenJDK 11

Are you planning to upgrade JVM in the near future? To which version?
We will moved on to the next LTS version. We are especially interested in the new GC and all the new stuffs.

Do you have constraints about this upgrade?
We have develop a JAVA-Object-to-Graph-Mapper and and a Java Agent that will need to be upgraded to Java Modules, but we must do it anyway someday. This artifact make all the magic behind.

About Javascript, we do not use it in ODB, so it's no matter.

**Reply by htu-un** [reply]
Do you run OrientDB embedded or stand-alone?
yes
    Which JVM version are you running?
jvm 11
    Are you planning to upgrade JVM in the near future? To which version?
Yes
    Do you have constraints about this upgrade?
no

**Comment by Eric24** [comment]
**Do you run OrientDB embedded or stand-alone?**
stand-alone

**Which JVM version are you running?**
JVM 8

**Are you planning to upgrade JVM in the near future? To which version?**
Yes planning to upgrade to JVM 15, but no set schedule

**Do you have constraints about this upgrade?**
No

**Comment by arch7tect** [comment]
Do you run OrientDB embedded or stand-alone?
Embedded

Which JVM version are you running?
1.8

Are you planning to upgrade JVM in the near future? To which version?
No

Do you have constraints about this upgrade?
we use many third-party libraries and some of them do not support Java versions greater than 1.8

**Comment by devent** [comment]
Do you run OrientDB embedded or stand-alone?
Embedded and Stand-alone.

Which JVM version are you running?
Java 11.

Are you planning to upgrade JVM in the near future? To which version?
Yes, next LTS.

Do you have constraints about this upgrade?
no

**Comment by ruudvring** [comment]
Do you run OrientDB embedded or stand-alone?
Embedded

Which JVM version are you running?
1.8

Are you planning to upgrade JVM in the near future? To which version?
No

Do you have constraints about this upgrade?
we use many third-party libraries and some of them do not support Java versions greater than 1.8

**Comment by PhantomYdn** [comment]
**Do you run OrientDB embedded or stand-alone?**
Embedded

**Which JVM version are you running?**
OpenJDK 8

**Are you planning to upgrade JVM in the near future? To which version?**
Not yet

**Do you have constraints about this upgrade?**
No

**Comment by luigidellaquila** [comment]
Thank you very much everybody, it helps a lot defining the strategy for the future.

It's clear that there are very different needs in the community and that Java 8 is still a constraint in some cases, but also an upgrade is very desirable in other cases.

For now we are experimenting to integrate GraalVM in the distribution for all the JVM versions (also Java 8), this will help us have a smooth transition.

I'll keep you posted, any further feedback is welcome of course!

Thanks

Luigi

**Reply by Solido** [reply]
+1 for Graal

**Comment by KamenIvanov** [comment]
Is there any information about transition of OrientDb to latest java versions? (16 or better 17 LTS?)

**Reply by luigidellaquila** [reply]
Hi @KamenIvanov 

We are now fixing latest small incompatibilities to make OrientDB run on Java 17. 
IMHO the best thing would be to release v 4.0 with native support to 17, but we still have to define the scope for backward compatibility

Thanks

Luigi

**Reply by KamenIvanov** [reply]
That's some great news! Thank you for replying, hope everything goes smooth :)

**Reply by nithyasharabheshwara** [reply]
> Hi @KamenIvanov
> 
> We are now fixing latest small incompatibilities to make OrientDB run on Java 17. IMHO the best thing would be to release v 4.0 with native support to 17, but we still have to define the scope for backward compatibility
> 
> Thanks
> 
> Luigi

Thanks. It is encouraging to know that jdk17 is being targeted for support. Old docs of orientdb 3 seemed to suggest to use only java 8, it got me worried.

**Reply by andrii0lomakin** [reply]
Hi guys, we already support JDK 17, to ensure that we do not break anything we have added build in GitHub actions.

**Reply by nithyasharabheshwara** [reply]
Is it a nightly ? Which version ?

**Reply by andrii0lomakin** [reply]
Yes, that is in the nightly builds of the 3.2 version.

**Comment by fbacchella** [comment]
Nashorn is still available: https://mvnrepository.com/artifact/org.openjdk.nashorn/nashorn-core
